pkg/netpol/eval/ingressanalyzer/ingress_analyzer.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2022
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ingressanalyzer
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+
+	ocroutev1 "github.com/openshift/api/route/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/np-guard/netpol-analyzer/pkg/netpol/eval"
+	"github.com/np-guard/netpol-analyzer/pkg/netpol/eval/internal/k8s"
+	"github.com/np-guard/netpol-analyzer/pkg/netpol/scan"
+)
+
+type (
+	IngressAnalyzer struct {
+		pe *eval.PolicyEngine // a struct type that includes the podsMap and
+		// some functionality on pods and namespaces which is required for ingress analyzing
+		servicesMap map[string]map[string]*k8s.Service // map from namespace to map from service name to its object
+		routesMap   map[string]map[string]*k8s.Route   // map from namespace to map from route name to its object
+		// k8sIngressMap map[string]map[string]*k8s.Ingress // map from namespace to map from ingress name to its object
+	}
+)
+
+// NewIngressAnalyzer returns a new IngressAnalyzer with an empty initial state
+func NewIngressAnalyzer() *IngressAnalyzer {
+	return &IngressAnalyzer{
+		pe:          eval.NewPolicyEngine(),
+		servicesMap: make(map[string]map[string]*k8s.Service),
+		routesMap:   make(map[string]map[string]*k8s.Route),
+	}
+}
+
+// NewIngressAnalyzerWithObjects returns a new IngressAnalyzer with relevant objects
+func NewIngressAnalyzerWithObjects(objects []scan.K8sObject) (*IngressAnalyzer, error) {
+	ia := NewIngressAnalyzer()
+	var err error
+	ia.pe, err = eval.NewPolicyEngineWithObjects(objects)
+	if err != nil {
+		return nil, err
+	}
+	for _, obj := range objects {
+		switch obj.Kind {
+		case scan.Service:
+			err = ia.upsertService(obj.Service)
+		case scan.Route:
+			err = ia.upsertRoute(obj.Route)
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+	return ia, err
+}
+
+// GetIngressControllerPod creates a new ingress-controller pod, as the source of all ingress connections
+func (ia *IngressAnalyzer) GetIngressControllerPod() eval.Peer {
+	ingressPod := &k8s.Pod{
+		Name:      "ingress-controller",
+		Namespace: "",
+	}
+	return &k8s.WorkloadPeer{Pod: ingressPod}
+}
```

Comment:
Add a const for the hard-coded fake name (and also add namespace).
Will also have to add this pod to the policy-engine, to evaluate permitted connectivity given an "unknwon" ingress controller pod.

Reply:
> Will also have to add this pod to the policy-engine, to evaluate permitted connectivity given an "unknwon" ingress controller pod.

do we already have a use case for this? should it be included in this PR ?

Reply:
not in this PR.. but in the next one